### PR TITLE
Add Authentication, Landing Page Modals, and Others

### LINKED ISSUE #15: Wire in Firebase Auth
I published [my life counter sample app](https://github.com/eeeeeric/life-counter-for-mtg) just now. It should serve as a good example on how to wire in Firebase Auth. The sample app is written using VueJS (and other components in the Vue ecosystem like vue-router and vuex).

The key things to do are:
* Add `firebase` and `firebaseui` dependencies
* Port [these changes](https://github.com/eeeeeric/life-counter-for-mtg/commit/3d47cb5317a0267c4e04cd698e796a753e93258e#diff-17803ee410d08d15bd5892bf6519f40a) which are the most important - it starts up the auth ui (this is a component provided by the `firebaseui` library and does all the hard work for you
* Port [these other changes](https://github.com/eeeeeric/life-counter-for-mtg/commit/3d47cb5317a0267c4e04cd698e796a753e93258e#diff-8cfead41d88ad47d44509a8ab0a109ad) which let you know when the auth state has changed (so you could perhaps persist the profile information somewhere)
* Store the user's profile information somewhere (I stored it using vuex - you might use redux) - you could also use a cookie
* Once you are authenticated, you don't really need to do anything. Calls to firebase using the firebase SDK will automatically include your auth token

### PR DESCRIPTION
**Pull request by acmorales**
A lot of odds and ends in this PR. Mostly groundwork for authentication and using the functions, can add more styles to the `LandingPage` and modals in a separate branch. Closes #15, #59, #60, and #61.
- Ability to login and `UserContext` which can be used to access the currently logged in user and `signIn()/signOut()` functions
- `withAuthentication()` can be used to require a user to be signed in to view a component
- Finally... the modals!
- Various helper functions in `helpers.js` and `api.js`
- Enabled cors in the functions
- Bumped up a major React (for Hooks) and Reactstrap (for Spinners) version
- Lightly re-styled `LandingPage`

**DEMO**
Must be signed in to view the board
withAuthentication

Modals
modals

### CODE REVIEW

**Comment by eeeeeric** [comment]
Do you have the latest changes from master in this branch? I'm hoping you don't and that's why CircleCI is failing.

File: firebase/functions/package.json [hunk 1]
```diff
@@ -13,13 +13,15 @@
...
-    "firebase-functions": "2.2.0",
-    "firebase-admin": "7.0.0",
     "@google-cloud/firestore": "0.19.0",
-    "uuid": "3.3.2",
-    "express": "4.16.4",
     "axios": "0.18.0",
-    "lodash": "4.17.11"
+    "cors": "^2.8.5",
+    "express": "4.16.4",
+    "firebase": "^5.8.5",
```
Comment (eeeeeric) [inline firebase/functions/package.json]: This actually looks to be a new dependency - is it necessary?

File: firebase/functions/test/index.test.ts [hunk 2]
```diff
@@ -100,7 +100,7 @@ describe('Cloud Functions Test Suite', function() {
     describe('Test importing a standard commander deck (1 commander, 99 other)', function() {
       describe('Import a standard commander deck for user1, via Cloud Function', function() {
         it('Make the HTTP call', async function() {
-          this.timeout(15000);
+          // this.timeout(15000);
```
Comment (eeeeeric) [inline firebase/functions/test/index.test.ts]: I would keep this timeout.

File: firebase/functions/package.json (hunk 1 above)
Comment (acmorales) [reply firebase/functions/package.json]: Good catch, must've added it while I was trying to run the functions. I'll remove it.

File: firebase/functions/test/index.test.ts (hunk 2 above)
Comment (acmorales) [reply firebase/functions/test/index.test.ts]: Also forgot to uncomment, will revert.

**Pull request merged**